modules/navier_stokes/src/fvbcs/WCNSFVInletTemperatureBC.C

```diff
@@ -0,0 +1,134 @@
+    _mdot_pp(isParamValid("mdot_pp") ? &getPostprocessorValue("mdot_pp") : nullptr),
+    _area_pp(isParamValid("area_pp") ? &getPostprocessorValue("area_pp") : nullptr),
+    _rho(isParamValid(NS::density) ? &getFunctor<ADReal>(NS::density) : nullptr),
+    _cp(isParamValid(NS::cp) ? &getFunctor<ADReal>(NS::cp) : nullptr)
+{
+  if (!dynamic_cast<INSFVEnergyVariable *>(&_var))
+    paramError(
+        "variable",
+        "The variable argument to WCNSFVInletTemperatureBC must be of type INSFVEnergyVariable");
+
+  // Density is often set as global parameters so it is not checked
+  if (_temperature_pp && (_velocity_pp || _mdot_pp || _energy_pp))
+    mooseWarning(
+        "If setting the temperature directly, no need for inlet velocity, mass flow or energy");
+
+  // Need enough information if trying to use a mass flow rate postprocessor
+  if (!_temperature_pp)
+  {
+    if (!_energy_pp)
+      mooseError("If not providing the temperature, the energy flow rate should be provided");
+    if (!_velocity_pp && !_mdot_pp)
+      mooseError("If not providing the inlet temperature, the inlet velocity or mass flow should "
+                 "be provided");
+    if (_velocity_pp && (!_rho || !_cp || !_area_pp))
+      mooseError("If providing the inlet velocity, the density, the area and the specific heat "
+                 "capacity should be provided as well");
+    if (_mdot_pp && !_cp)
+      mooseError("If providing the inlet mass flow rate, the inlet specific heat capacity should "
+                 "be provided as well");
+  }
+  else if (params.isParamSetByUser("scaling_factor"))
+    paramError("scaling_factor",
+               "The scaling factor is meant to adjust for a different area or "
+               "mass flow rate, it should not be set if the temperature is set directly");
+}
+
+Real
+WCNSFVInletTemperatureBC::boundaryValue(const FaceInfo & fi) const
+{
+  if (_area_pp)
+    if (MooseUtils::absoluteFuzzyEqual(*_area_pp, 0))
+      mooseError("Surface area is 0");
+
+  if (_temperature_pp)
+    return *_temperature_pp;
+  else if (_velocity_pp)
+  {
+    ADReal rho;
+    ADReal cp;
+
+    // TODO: Make this a utility to all FVBCs
+    const bool use_elem = fi.faceType(_var.name()) == FaceInfo::VarFaceNeighbors::ELEM;
+
+    if (use_elem)
+    {
+      rho = (*_rho)(std::make_tuple(
+          &fi, Moose::FV::LimiterType::CentralDifference, true, fi.elem().subdomain_id()));
+      cp = (*_cp)(std::make_tuple(
+          &fi, Moose::FV::LimiterType::CentralDifference, true, fi.elem().subdomain_id()));
+    }
+    else
+    {
+      rho = (*_rho)(std::make_tuple(
+          &fi, Moose::FV::LimiterType::CentralDifference, true, fi.neighborPtr()->subdomain_id()));
+      cp = (*_cp)(std::make_tuple(
+          &fi, Moose::FV::LimiterType::CentralDifference, true, fi.neighborPtr()->subdomain_id()));
+    }
+
+    return _scaling_factor * (*_energy_pp) / (*_area_pp * rho.value() * *_velocity_pp * cp.value());
```

Comment:
This is what `singleSidedFaceArg` does. I would suggest moving that method and the `_face_info` member down to `FVBoundaryCondition`, set `_face_info = &fi` in this method, and then call `singleSidedFaceArg`. Similarly for all the other times this is done in this PR

Reply:
true true. it s what I used for the flux BCs. 

Any reason the DirichletBCs are Real and not ADReal?

Reply:
In the beginning we only imagined fairly simple Dirichlet conditions: e.g. constants and direct functions of space and time. But now if you have more advanced conditions that could involve another variables nonlinear degrees of freedom, then we could change it to return `ADReal` for sure

Reply:
so I remember why I did not do that now. computeValue is const, so I cant assign _face_info.

Reply:
so either we
- keep it as I have
- assign _face_info earlier, like before computingValue (well I think i like that most)
- make all computeValue not const

Reply:
Moving up only singleSidedFaceArgs to FVBoundaryCondition feels like all these routines should move together.
Duplicating it into FVDirichletBCBase isnt great either though

Reply:
or we could just add a non-const version of boundaryValue. Seems to defeat the purpose of const-ness most of the time though.

Reply:
What are you including in "all these routines"? I'm a fan of moving `singleSidedFaceArg` up to `FVBoundaryCondition` and then adding a default `const FaceInfo * face_info = nullptr` argument that if `nullptr` automatically resolves to `_face_info`. This is what is done in `FVFluxKernel::faceArgSubdomains` (well at least on the interpolator branch but I think in upstream as well)

Reply:
So I guess that also would necessitate moving `_face_info` up to `FVBoundaryCondition` as well (which seems a natural move to me)

Reply:
> so I remember why I did not do that now. boundaryValue is const, so I cant assign _face_info.

This is the problem really. Moving attributes is ok imo.

Reply:
You don't need `_face_info` assigned if you add a `face_info` parameter to `singleSidedFaceArg`

Reply:
sounds good